Add new settings page and element positioning preview for goggle OSD

Nikolas-S opened the pull request:
**WARNING: The way settings are saved was modified on the main branch recently. If you test this PR and change the visibility of OSD elements, you will have to reset your settings using the reset button on the Firmware page just *before* going back to 9.0.12, otherwise your OSD elements may be invisible in 9.0.12.**

Created a new settings page for the goggle OSD.
Moved the osd mode setting from image settings to the new osd settings.

The new page allows for individual positioning and visibility settings for every goggle osd element.
The element position can be set separately for 4x3 and 16x9 modes, visibility is shared between modes.
The element position is set using x and y sliders. I think this is ok for now but the goal is to have the elements show and update live, similar to the image settings. (don't have my goggles yet and the osd seemed wonky on the emulator)

<img width="779" alt="osd_page" src="https://user-images.githubusercontent.com/35898034/232610267-72b0b161-1203-4c05-a2f2-70111a846c45.png">

I also added acceleration when moving the sliders to make positioning less tedious while not losing precision, would love some feedback on how it feels on real goggles.
If anyone has ideas for improvements, let me know.

For the devs: 
I added a new app state APP_STATE_SUBMENU_ITEM_FOCUSED.
When in that state, a pages on_roller is called but without automatically changing the submenu item selection, allowing the page to correctly handle input for items like dropdown menus and sliders.
I propose we use the new state for future changes, seems cleaner than separate app states for every element or other workarounds.

SumolX commented:
Very Nice! May I make a few recommendations:

1. Rename "Goggle OSD" -> OSD?
2. Highly recommend that this PR does include the live view functionality though to preview the modification and I believe this should work in the emulator for you to try out.
3. Perhaps add a Save Settings option (Similar to Update Clock) just in case you want to play with the values but aren't sure just yet if you want them saved.

Overall I really like the layout and excellent use of the dropdown.

Nikolas-S commented:
> Very Nice! May I make a few recommendations:
> 
>     1. Rename "Goggle OSD" -> OSD?
> 
>     2. Highly recommend that this PR does include the live view functionality to preview the modification as we are just guessing.  This should work in the emulator.
> 
>     3. Perhaps add a Save Settings option (Similar to Update Clock) just in case you want to play with the values but aren't sure just yet if you want them saved.
> 
> 
> Overall I really like the layout and excellent use of the dropdown.
> 
> As for they wonkyness I believe the main window is 1080p and the live view is 720p and thus it draws the live view within that area starting from the top left.... so for you testing purpose... maybe set the main window to 720p instead to ensure its going to display properly. I can test it out for you on the goggles.

1. Was going with goggle osd so that no one expects any magical betaflight integration but it's pretty obvious what the page does so might change it to OSD
2. I took another look at the image settings, I think I understand what's going on, missed the part where it switches to 720p.
3. With the blind sliders I didn't think it necessary but with a preview screen that's a good idea.

My goggles are already in the mail and should arrive later this week, I'll convert this pr to a draft for now and see if I can come up with a preview in the mean time. I'll let you know if i got something to test 😄

Nikolas-S commented:
Added a preview for the element positioning, which is accessible from the OSD settings page.
The user can change the visibility/position of individual elements with a live preview of the changes.
The position can be set separately for both modes, while visibility is shared.
Resetting the elements resets positions for both modes, as well as visibility.
The OSD mode is accessible from both the OSD settings page for quick access and the preview UI.

I also added acceleration when moving the sliders to make positioning less tedious without losing precision. I would love some feedback on how it feels on real goggles.
If anyone has ideas for improvements, let me know.

Fair amount of new stuff with this one, so the more people hunting for bugs, the better.

![image](https://user-images.githubusercontent.com/35898034/233216423-0b8aecb3-3ed9-4507-801f-c29e339da391.png)

![image](https://user-images.githubusercontent.com/35898034/233216244-c1893c26-4c55-400c-a98b-9365619c3880.png)

SumolX commented:
Tested on my goggles and it works very well.  The acceleration took some getting used to as when i tried slowing down to fine tune the position the movement still had a significant skip.  I do have a question... without looking at the code are you handling 1080p live mode because I noticed the footnote stating 720p canvas?  Otherwise I love it!

Nikolas-S commented:
Changed the slider scroll speed to depend on the time between inputs, aka how fast you spin the wheel.
Also rebased to use the new osd_resource_path function.

SumolX commented:
> Changed the slider scroll speed to depend on the time between inputs, aka how fast you spin the wheel. Also rebased to use the new osd_resource_path function.

The scrolling is very good now.  I can transition from rapid to fine movement nicely. Overall it i think it works great!

pitts-mo commented:
Just toggled on and off elements to test. Looks nice. 

Suggestions for this PR:
-Somehow default all elements ON.
-Add a note about to the positioning screen eg: "Note: This interface will display all OSD elements to allow for positioning. However, some OSD elements will not persist during normal operation." 

I was a little confused by the persistence of HDZero VRX elements while configuring OSD in analog mode :-)

Looking forward to goggle battery voltage display :-)

bkleiner commented on src/ui/ui_osd_element_pos.c:
```diff
@@ -0,0 +1,665 @@
+
+// ###########################################################################
+// ########################## local variables start ##########################
+// ###########################################################################
```
_imho_ blocks like these have little utility.

Nikolas-S replied:
i suppose it's pretty obvious when looking at the code, removed the blocks.

Nikolas-S commented:
> Just toggled on and off elements to test. Looks nice.
> 
> Suggestions for this PR: -Somehow default all elements ON. -Add a note about to the positioning screen eg: "Note: This interface will display all OSD elements to allow for positioning. However, some OSD elements will not persist during normal operation."
> 
> I was a little confused by the persistence of HDZero VRX elements while configuring OSD in analog mode :-)
> 
> Looking forward to goggle battery voltage display :-)

Not sure what you mean by default all elements ON, could you elaborate?
Improved the user hint in the latest commit.

Nikolas-S commented:
> Hmm, Seems like we need to handle or prevent the long press that returns you to the main menu.
> 
> You can return to the main goggle menu and leave the Adjust OSD elements menu stuck in the display. if you go back into Adjust OSD elements you can then cancel to clear it from the display.

Nice find! A long press should now discard any changes and return to the main menu.
Probably noteworthy that a long press in the image settings preview returns to the main menu while _saving_ all changes, but I think saving settings with a long press is not the best idea since it can happen accidentally and seems kinda unintuitive, so maybe worth a refactor at some point.

pitts-mo commented:
> > Just toggled on and off elements to test. Looks nice.
> > Suggestions for this PR: -Somehow default all elements ON. -Add a note about to the positioning screen eg: "Note: This interface will display all OSD elements to allow for positioning. However, some OSD elements will not persist during normal operation."
> > I was a little confused by the persistence of HDZero VRX elements while configuring OSD in analog mode :-)
> > Looking forward to goggle battery voltage display :-)
> 
> Not sure what you mean by default all elements ON, could you elaborate? Improved the user hint in the latest commit.

Maybe something with my usage caused all my elements to show as OFF?

I had been using 16x9 OSD and they were all showing in 9.0.12 release.
When I tried a master build I saw the OSD elements only display briefly (<1 second?) when starting into analog mode.
When I tried a build from this PR I saw the same behavior and on entering the Adjust OSD elements menu I found they were all off.

Nikolas-S commented:
> > > Just toggled on and off elements to test. Looks nice.
> > > Suggestions for this PR: -Somehow default all elements ON. -Add a note about to the positioning screen eg: "Note: This interface will display all OSD elements to allow for positioning. However, some OSD elements will not persist during normal operation."
> > > I was a little confused by the persistence of HDZero VRX elements while configuring OSD in analog mode :-)
> > > Looking forward to goggle battery voltage display :-)
> > 
> > 
> > Not sure what you mean by default all elements ON, could you elaborate? Improved the user hint in the latest commit.
> 
> Maybe something with my usage caused all my elements to show as OFF?
> 
> I had been using 16x9 OSD and they were all showing in 9.0.12 release. When I tried a master build I saw the OSD elements only display briefly (<1 second?) when starting into analog mode. When I tried a build from this PR I saw the same behavior and on entering the Adjust OSD elements menu I found they were all off.

That is interesting, perhaps it has something to do with my recent pr where the way bools are saved was changed? (https://github.com/hd-zero/hdzero-goggle/pull/196)
I don't know if the settings file is reset during updates, if your settings file still has them saved as 0/1, but the goggle expects true/false that might explain it.
If you change the visibility and save, does it retain the configuration after reboot?

This is from just before the pr: https://github.com/hd-zero/hdzero-goggle/actions/runs/4685509451
This is from just after the pr: https://github.com/hd-zero/hdzero-goggle/actions/runs/4717254064

Nikolas-S commented:
@pitts-mo regarding your first two points, I created issue #236 which addresses them, those are important to get resolved before a next release.

Regarding your OSD being gone after going back to 9.0.12, I am not sure what exactly 9.0.12 is based off of but ever since #182 from late February, the visibility of OSD elements has been saved as 0/1 in the settings, just without any UI for the user to change the settings. So if pr #182 is included in 9.0.12, testing this new pr would have made your OSD element visibility settings incompatible with 9.0.12, since bools are now saved as true/false instead of 0/1.

Other than that, if you stay only within the new pr, are settings being retained correctly? 
And were there any other issues you encountered during testing?

pitts-mo commented:
Sorry @Nikolas-S , I was away for a bit.

Yes, My observations appear to agree. And Yes, So far my settings have been working as expected even when switching between this PR and a recent build from master.
Thank you,
-P

Nikolas-S commented:
Rebased after recent OSD element changes.

Nikolas-S commented:
If there are no more ideas for improvements/known issues, think we could get this merged?
If so, we should either add a warning to the readme that changing settings will cause issues when going back to 9.0.12, or temporarily hide the OSD settings page by commenting out pp_osd from page_packs until the next release, so people testing won't accidentally make their settings incompatible with 9.0.12. (since there's no easy way to reset settings in 9.0.12)

Dom4n commented:
This PR is just awesome! Thank You very much for this! Now I can set antenna indicators so it makes sense for me :smile:  
I hope it will be approved and merged soon. 
PR #249 is working as expected too, just tested it twice (hop between old and new firmware).

Now I`m on fw from this build: https://github.com/Nikolas-S/hdzero-goggle/actions/runs/5054980978 and it works as should. Flew today five packs and it was rock solid.

Nikolas-S commented:
> This PR is just awesome! Thank You very much for this! Now I can set antenna indicators so it makes sense for me :smile:  
> I hope it will be approved and merged soon. 
> PR #249 is working as expected too, just tested it twice (hop between old and new firmware).
> 
> Now I`m on fw from this build: https://github.com/Nikolas-S/hdzero-goggle/actions/runs/5054980978 and it works as should. Flew today five packs and it was rock solid.

Thank you, glad you like it :)

Nice, just don't risk too much when flying PRs 😁

Dom4n commented:
Feature request: import/export settings to/from SD card would be really useful.
